src/crypto/PSAKeyAllocator.h

```diff
@@ -0,0 +1,212 @@
+
+/**
+ * @brief Calculates PSA key ID for Node Operational Certificate private key for the given fabric.
+ */
+constexpr psa_key_id_t MakeOperationalKeyId(FabricIndex fabricIndex)
+{
+    return to_underlying(KeyIdBase::Operational) + static_cast<psa_key_id_t>(fabricIndex);
+}
+
+/**
+ * @brief Interface for PSA key allocation.
+ *
+ * The PSA Key Allocator interface provides an abstraction that allows the application to
+ * allocate PSA keys in a secure environment. This class uses a concept that isolates the
+ * application from the actual key material. The secure location may vary depending on the
+ * cryptographic hardware used. Using this class a platform can implement this interface to
+ * allocate keys in the specific secure location.
+ *
+ * In some cases key attributes must be redefined to match the specific requirements of the
+ * secure location and the cryptographic hardware.
+ *
+ * This class keeps the static instance and uses it in the all the places where the key
+ * should be persisted. You can add the usage of this class anywhere in the code by calling the
+ * GetPSAKeyAllocator() function.
+ *
+ * If the static instance is not set, the default implementation is used.
+ *
+ * To change the static instance of the PSAKeyAllocator, you can call the SetPSAKeyAllocator function.
+ */
+class PSAKeyAllocator
+{
+public:
+    /**
+     * @brief Destructor for PSAKeyAllocator.
+     */
+    virtual ~PSAKeyAllocator() = default;
+
+    /**
+     * @brief Get the Device Attestation Key (DAC) ID.
+     *
+     * @return psa_key_id_t The DAC key ID.
+     */
+    virtual psa_key_id_t GetDacKeyId() = 0;
+
+    /**
+     * @brief Get the Node Operational Certificate key ID for a given fabric index.
+     *
+     * @param fabricIndex The fabric index for which the operational key ID is requested.
+     * @return psa_key_id_t The operational key ID.
+     */
+    virtual psa_key_id_t GetOpKeyId(FabricIndex fabricIndex) = 0;
+
+    /**
+     * @brief Allocate a new Intermittently Connected Devices (ICD) key ID.
+     *
+     * ICD uses two keys - AES CCM and HMAC(SHA 256).
+     * Each key must be allocated separately.
+     *
+     * @return psa_key_id_t The newly allocated ICD key ID.
+     */
+    virtual psa_key_id_t AllocateICDKeyId() = 0;
+
+    /**
+     * @brief Update the key attributes before storing the key.
+     *
+     * In some cases the key attributes must be redefined to match the specific requirements of the
+     * secure location and the cryptographic hardware. This method allows the platform to update the
+     * key attributes before storing the key.
+     *
+     * Read the current key attributes to determine the key type, algorithm, and usage flags. Update
+     * the key attributes as needed.
+     *
+     * @param attrs Pointer to the key attributes structure to be updated.
+     */
+    virtual void UpdateKeyAttributes(psa_key_attributes_t & attrs) = 0;
+
```

Comment:
Out of scope of this PR: I do understand that Group Keys are not stored in non-volatile secure storage currently?

Reply:
AFAIK, they're stored in settings.